mitosheet/src/components/LoadingIndicator.tsx

```diff
@@ -1,43 +1,275 @@
+    useEffect(() => {
+        if (props.loading.length === 0) {
+            setDisplay(false);
+        } else if (props.loading.length > 0) {
+            const timeout = setTimeout(() => {
+                setDisplay(true);
+            }, 500);
+            return () => {clearTimeout(timeout)}
+        }
+    }, [props.loading.length]);
+
+    // This effect tracks the top message and when it becomes the top message
+    // so that we can give the user special messages when things have been 
+    // loading for longer than 10 seconds
     useEffect(() => {
-        setTimeout(() => {
-            setDisplay(true);
-        }, 500);
-    }, []);
-
-    // We start the indicator at -1, so that we don't display anything
-    // for the first half second. This makes us only display the indicator
-    // for actually long running operations.
-    if (!display) {
+        const interval = setInterval(() => {
+            const messagesToDisplay = getMessageTypesToDisplay(props.loading);
+            if (messagesToDisplay.length === 0) {
+                setCurrentLoadingMessage(undefined);
+            } else {
+                setCurrentLoadingMessage(prevLoadingMessage => {
+                    const topMessageID = messagesToDisplay[0][1];
+                    if (prevLoadingMessage === undefined || topMessageID !== prevLoadingMessage[1]) {
+                        return [Date.now(), topMessageID];
+                    }
+                    return prevLoadingMessage;
+                })
+            }
+            // We always refresh this state, though, so that this rerenders 
+            // so that we can refresh the messaging to the user
+            setCurrentLoadingMessage(prevCurrentLoadingMessage => {
+                if (prevCurrentLoadingMessage === undefined) return prevCurrentLoadingMessage;
+                return [prevCurrentLoadingMessage[0], prevCurrentLoadingMessage[1]];
+            })
+        }, 1000)
+        return () => {clearInterval(interval)};
+    }, [props.loading])
+
+
+    const messagesToDisplay = getMessageTypesToDisplay(props.loading);
+
+    if (!display || messagesToDisplay.length === 0) {
         return <React.Fragment/>
     }
 
     return (
         <div className='loading-indicator-container'>
-            <p className='ml-20px'>
-                Loading<LoadingDots/>
+            <p className='loading-indicator-header text-header-3 text-color-white-important'>
+                Processing {messagesToDisplay.length} edit{messagesToDisplay.length <= 1 ? '' : 's'}
             </p>
+            <div className='loading-indicator-content'>
+                {messagesToDisplay.map((([messageType, message_id], index) => {
+                    const is10SecondsAgo = currentLoadingMessage && message_id === currentLoadingMessage[1] && (currentLoadingMessage[0] <= Date.now() - 10 * 1000);
+
+                    return (messageType !== undefined && 
+                        <div 
```

Comment:
The font is not vertically centered with the other loading indicator and icon

<img width="707" alt="Screen Shot 2022-06-20 at 11 32 32 PM" src="https://user-images.githubusercontent.com/18709905/174710594-d93ba22d-ea1e-46f4-a376-25de146b2e47.png">